src/main/java/com/google/cloud/spanner/pgadapter/statements/IntermediatePreparedStatement.java

```diff
@@ -159,34 +176,283 @@ public IntermediatePortalStatement bind(
+          || Arrays.stream(this.parameterDataTypes).anyMatch(p -> p == 0)) {
+        // Note: We are only asking the backend to parse the types if there is at least one
+        // parameter
+        // with unspecified type. Otherwise, we will rely on the types given in PARSE.
+
+        // Transform the statement into a select statement that selects the parameters, and then
+        // extract the types from the result set metadata.
+        Statement selectParamsStatement = transformToSelectParams(parameters);
+        if (selectParamsStatement == null) {
+          // The transformation failed. Just rely on the types given in the PARSE message. If the
+          // transformation failed because the statement was malformed, the backend will catch that
+          // at
+          // a later stage.
+          describeFailed = true;
+          ensureParameterLength(parameters.size());
+        } else {
+          try (ResultSet paramsResultSet =
+              connection.analyzeQuery(selectParamsStatement, QueryAnalyzeMode.PLAN)) {
+            extractParameterTypes(paramsResultSet);
+          } catch (SpannerException exception) {
+            // Ignore here and rely on the types given in PARSE.
+            describeFailed = true;
+            ensureParameterLength(parameters.size());
+          }
+        }
+      }
+      if (columnsResultSet != null) {
+        return new DescribeStatementMetadata(this.parameterDataTypes, columnsResultSet);
+      }
+
+      if (this.parsedStatement.isUpdate()
+          && (describeFailed || !Strings.isNullOrEmpty(this.name))) {
+        // Let the backend analyze the statement if it is a named prepared statement or if the query
+        // that was used to determine the parameter types failed, so we can return a reasonable
+        // error
+        // message if the statement is invalid. If it is the unnamed statement or getting the param
+        // types succeeded, we will let the following EXECUTE message handle that, instead of
+        // sending
+        // the statement twice to the backend.
+        connection.analyzeUpdate(
+            Statement.of(this.parsedStatement.getSqlWithoutComments()), QueryAnalyzeMode.PLAN);
+      }
+      return new DescribeStatementMetadata(this.parameterDataTypes, null);
+    } finally {
+      if (columnsResultSet != null) {
+        columnsResultSet.close();
+      }
+    }
+  }
+
+  @VisibleForTesting
+  Statement transformToSelectParams(Set<String> parameters) {
+    switch (this.parsedStatement.getType()) {
+      case QUERY:
+        return transformSelectToSelectParams(parameters);
+      case UPDATE:
+        return transformDmlToSelectParams(parameters);
+      case CLIENT_SIDE:
+      case DDL:
+      case UNKNOWN:
+      default:
+        return Statement.of(this.parsedStatement.getSqlWithoutComments());
+    }
+  }
+
+  private Statement transformSelectToSelectParams(Set<String> parameters) {
```

Comment:
Can we add examples as comment? Same for other parsers

Reply:
Done